### PR TITLE
hcm: support raw header value

### PR DESCRIPTION
**Pull request by kyessenov**
Commit Message: Implement base64 encoding for dynamic raw header values.
Risk Level: low
Testing: done
Docs Changes: yes 
Release Notes: yes
Issue: https://github.com/envoyproxy/envoy/issues/29590

### CODE REVIEW

**Comment by adisuissa** [comment]
Assigning @yanjunxiang-google for a first pass.
/assign @yanjunxiang-google

File: source/common/router/header_parser.cc [hunk 1]
```diff
@@ -37,16 +37,21 @@ parseHttpHeaderFormatter(const envoy::config::core::v3::HeaderValue& header_valu
   if (!Http::HeaderUtility::isModifiableHeader(key)) {
     throw EnvoyException(":-prefixed or host headers may not be modified");
   }
+  const bool raw_value = !header_value.raw_value().empty();
+  if (raw_value && !header_value.value().empty()) {
+    throw EnvoyException("cannot specify both value and raw_value");
```
Comment (yanjunxiang-google) [inline source/common/router/header_parser.cc:42]: Just a thought: If both raw_value and value is configured, should we just take priority on raw_value and ignore value? Is that more tolerable for mis-configure?

**Comment by yanjunxiang-google** [comment]
LGTM in general

@yanavlasov Could you please take a look?

File: source/common/router/header_parser.cc (hunk 1 above)
Comment (kyessenov) [reply source/common/router/header_parser.cc:42]: From the proto spec, raw_value sounded like a one-of promotion. I think it's better to dis-allow ambiguous cases and reject, we can always relax the rejection later.

**Comment by kyessenov** [comment]
Base64 is commonplace (e.g. JWT), and lossy encoding is not always desired if the payload needs to be decoded. I was considering a formatter function, but I think we don't actually have any yet, and I'm worried about the hand-written parser code in the formatter. CEL might be another option, we can expose a Base64 encoding function and require CEL formatter extension.  In any case, what do we actually `raw_value` to mean in HCM context?

**Comment by yanavlasov** [comment]
I think base64 encoding the value will lock this feature into a very specific use case.

I think the use case is the following: I want to add a header with the `0xC9` character in it. I need to use the `raw_value` to avoid UTF-8 check in proto string.

If Envoy also base64 encodes it it makes the feature too limited. Which is I would prefer to add an operator for base64 encoding. I think it will be more challenging to implement base64 operator since we do not have support for recursive commands, but I would rather go that way, to avoid locking raw_value usage into specific behavior.

/wait-any

**Comment by kyessenov** [comment]
Let me explore CEL option, as it is more robust to adding new custom functions.

**Pull request closed without merging**